Don't bypass tagging interface

tophmatthews opened the pull request:
Ref #23738

tophmatthews committed:
update ad_abs_ref test. Ref #23738

tophmatthews commented:
I'm taking another hack at trying to get ADKernel derived objects to accumulate the absolute value of the residual. This is related to #23399 which works for Kernel derived objects. In `Kernel`, `TaggingInterface::accumulateTaggedLocalResidual()` is used to accumulate the absolute value of the contributions of the residual and adds it to a custom tag. `ADKernel` uses it's own system for accumulating residuals and adding them to tags...and I'm getting lost in where to plug this all in. Here's what I've tried and found out:
- I tried separating out the method where the absolute value is accumulated in `TaggingInterface::accumulateAbsolueValueLocalResidual`. 
- Since the residual is accumulated as a `std::vector<Real>` in `ADKernel` instead of `DenseVector<Number>`, I am able to make an overloaded version of this method. 
- This method hits a seg fault in the AD version (the one with `std::vector<Real>`) iterating through the absolute value residual blocks (`_absre_blocks`)
- `_absre_blocks `is defined in `TaggingInterface.h` as `std::vector<DenseVector<Number> *>`
- The size of the vector is set by the number of absolute vector tags (usually one).
- Each element of the vector is set by `_absre_blocks[i] = &assembly.residualBlock(ivar, tag._type_id);`
- ADKernel objects do not have residualBlocks (I think) or at least not in the same manner, hence the seg fault

With Fande gone (who implemented the tagging system) I'm hoping @lindsayad you may have insight into this? Help!

tophmatthews commented:
I added a test here that can be used to verify it works. `test/tests/problems/reference_residual_problem/abs_ref.i` is the non-AD version of `ad_abs_ref.i`. `ad_abs_ref.i` only has the BodyForce terms as AD. These two inputs can be run for the first time step. The reference residual value for each should be exactly the same (u_ref, v_ref) but are not, for example...
```
 1 Nonlinear |R| = 8.895740e-15
   Solution, reference convergence variable norms:
   u: 2.35e-16  u_ref: 7.02e-01  (3.34e-16)
   v: 8.89e-15  v_ref: 1.40e-06  (6.33e-09)
Nonlinear solve converged due to CONVERGED_FNORM_RELATIVE iterations 1
``

tophmatthews committed:
attempt at accumulating abs value of residual in ADKernel

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1487877/) on 14fb45e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24206/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24206/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 6051fa3507fbebb9fa4c6d475fddca1bde72d6c4`

lindsayad commented:
> With Fande gone (who implemented the tagging system) I'm hoping @lindsayad you may have insight into this? Help!

😄 I'll add this towards the top of my task list

lindsayad committed:
Allow ADKernel objects to leverage the absolute value tagging machinery

And also maintain accumulating residuals on a per element per ADKernel
basis to in order to not fail computeResidualAndJacobian simulations
in dbg mode (e.g. we can't have the linear residual evaluated on a
per element basis while the nonlinear residual (computed via
`computeResidualAndJacobian) is computed on a per element per residual
objects basis else we will fail the check in libmesh where we verify
equivalent residual evaluations with no tolerancing)

lindsayad commented:
Ok @tophmatthews this appears to make things work

lindsayad commented:
If you agree, let's mark this as ready to review

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1503644/) on 4bc0bd4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24206/site)

This comment will be updated on new commits.

lindsayad committed:
Refactor such that Assembly::processXXXX only called by TaggingInterface

And consolidate some of the Assembly APIs

tophmatthews commented:
Hm, close, but doesn't match the tag values between AD and non-AD exactly. Let me dig a bit...

tophmatthews commented:
Ah whoops, I was looking at the wrong values, yes this fixed it!

lindsayad commented:
Ok @tophmatthews I'm actually going to keep this in draft because I'm doing a larger refactor to ensure people (including myself) don't bypass the TaggingInterface in the future

lindsayad commented:
@roystgnr this is the refactor I mentioned this morning

tophmatthews commented:
Sounds good @lindsayad , I updated the test that was in there to be full AD.

lindsayad commented:
Oh oops. I just force pushed. Can you rebase your update to full AD on top of what I just pushed, and then push?

tophmatthews committed:
fully adify ad_abs_ref

lindsayad committed:
See if this satisfies clang

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1505385/) on 555407e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>0dd4ff</th>
      <th colspan=4>#24206 555407</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.17%</td>
      <td>85.21%</td>
      <td>+0.04%</td>
      <td>94.37%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>87143</td>
      <td>87155</td>
      <td>+12</td>
      <td>201</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15179</td>
      <td>15132</td>
      <td>-47</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>0dd4ff</th>
      <th colspan=4>#24206 555407</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.13%</td>
      <td>90.22%</td>
      <td>+0.09%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3535</td>
      <td>3571</td>
      <td>+36</td>
      <td>65</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>387</td>
      <td>387</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage/contact)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>0dd4ff</th>
      <th colspan=4>#24206 555407</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.60%</td>
      <td>83.62%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10767</td>
      <td>10781</td>
      <td>+14</td>
      <td>22</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2112</td>
      <td>2112</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage/navier_stokes)

### Ray tracing

<table>
  <thead>
    <tr>
      <th></th>
      <th>0dd4ff</th>
      <th colspan=4>#24206 555407</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>93.77%</td>
      <td>93.77%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4080</td>
      <td>4079</td>
      <td>-1</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>271</td>
      <td>271</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage_diff/ray_tracing)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24206/coverage/ray_tracing)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24206/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Make a lot of stuff in TaggingInterface private

lindsayad committed:
Add capability in framework for filtering reference residual tag

This allows us to remove the now-private data access in tensor
mechanics `GeneralizedPlaneStrain`

lindsayad committed:
Remove unused declaration

lindsayad committed:
Refactor slip BC classes

And add a setResidual overload in TaggingInterface that takes
a lambda that can operate on the residuals corresponding to the
`_vector_tags`

lindsayad committed:
Fix tag filtering for ReferenceResidualProblem

lindsayad committed:
processResidual -> addResidual

lindsayad committed:
processJacobian -> addJacobian

lindsayad committed:
Change relevant Assembly routine names from addXXX to cacheXXX

I want the user-facing tagging interfaces to be addXXX and setXXX to mimic the
add and set from libMesh/PETSc vectors/matrices, but I want the
Assembly implementation names to reflect what is actually happening
behind the scenes

lindsayad committed:
Make Assembly tag related methods private

At least as many as we can. Ones with default arguments can't be made
private (at least not with minimial effort) as they are used by the
constraint system. I remember Fande spending a lot of effort trying to
get the constraint system to work with tagging and he eventually gave
up.

Also delete some deprecated APIs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1511431/) on f2cc786 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24206/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24206/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97e1f3db58c8f7ebfde45b5aaf7b292e0584e173`

lindsayad committed:
clang-format

lindsayad committed:
Updating modules to leverage TaggingInterface post Assembly lockdown

lindsayad committed:
Avoid dynamic allocation of dof indices

lindsayad committed:
Make nonlocal classes use tagging interface

lindsayad committed:
Adapt NodeFaceConstraint Jacobians to TaggingInterface

lindsayad committed:
Adapt NodeFaceConstraint residuals to TaggingInterface

lindsayad committed:
Use TaggingInterface in NonlinearSystemBase

This removes the last non-blessed friend from Assembly::LocalDataKey

But this also introduced a somewhat undesirable friend in TaggingInterface.
I should probably look into that due to my neuroticism

lindsayad committed:
Update ElemElemConstraint and NodeElemConstraint to TaggingInterface

lindsayad committed:
Update NodalConstraint Jacobian to use TaggingInterface

lindsayad committed:
Update NodalConstraint residual to use TaggingInterface

lindsayad committed:
No more default tag arguments

lindsayad committed:
Make for slightly less compiler errors in relap-7/sam

lindsayad commented:
This should be ready for review. I've merged fixes for all apps other than SAM. SAM will require some of the new APIs in this PR to complete its update, but I have that patch ready to go

lindsayad commented:
@tophmatthews I'm sorry that I took over this PR. Is there anything else you wanted to add?

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1519077/) on c7aaed4 : invalidated by @lindsayad

moosebuild commented:
Job [Apptainer controlled apps](https://civet.inl.gov/job/1519055/) on c7aaed4 : invalidated by @lindsayad

moosebuild commented:
Job [Controlled app tests](https://civet.inl.gov/job/1519078/) on c7aaed4 : invalidated by @lindsayad

moosebuild commented:
Job [External app tests](https://civet.inl.gov/job/1519081/) on c7aaed4 : invalidated by @lindsayad

lindsayad commented:
https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/741 for SAM

lindsayad commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -44,18 +51,54 @@ class TaggingInterface
+  void prepareVectorTag(Assembly & assembly,
+                        unsigned int ivar,
+                        const ReferenceResidualProblem * ref_problem,
+                        bool prepare_non_ref_tags);
```
Add doxygen. Also maybe either
1. change to `bool prepare_ref_tags` (no negation)
2. Or make this an enum

loganharbour commented on modules/contact/src/constraints/ComputeFrictionalForceCartesianLMMechanicalContact.C:
```diff
@@ -327,24 +331,27 @@ ComputeFrictionalForceCartesianLMMechanicalContact::enforceConstraintOnDof(
-      (component_normal == 0 || component_normal == 2) ? dof_index_y : dof_index_x,
-      _vector_tags,
-      _matrix_tags);
+  addResidualsAndJacobian(
```
Forming the dof indices and scaling factors as an array would clean up all of the ternary logic.

Just a suggestion, I suspect that you're very much over this PR :-)

lindsayad replied:
Can you show what you're thinking? I don't get it yet.

loganharbour replied:
Not relevant to this PR; let's get this in instead.

loganharbour commented on modules/navier_stokes/include/bcs/MomentumFreeSlipBC.h:
```diff
@@ -23,9 +23,10 @@ class MomentumFreeSlipBC : public NodalNormalBC
 
 protected:
-  virtual Real computeQpResidual() override;
+  Real computeQpResidual() override { mooseError("Function unused"); }
```
We probably need a macro or something of the sort for this. next cleanup PR 😆

loganharbour commented on modules/tensor_mechanics/src/kernels/GeneralizedPlaneStrainOffDiag.C:
```diff
@@ -132,20 +131,18 @@ GeneralizedPlaneStrainOffDiag::computeDispOffDiagJacobianScalar(unsigned int com
+                                                   component) *
+                               _grad_test[_i][_qp](component);
+    accumulateTaggedLocalMatrix();
+    const auto ke_copy = _local_ke;
+
+    prepareMatrixTag(_assembly, jvar, _var.number());
+    ke_copy.get_transpose(_local_ke);
```
Member variable to avoid dense matrix allocations?

loganharbour commented on modules/tensor_mechanics/src/scalarkernels/GeneralizedPlaneStrain.C:
```diff
@@ -40,53 +40,26 @@ GeneralizedPlaneStrain::GeneralizedPlaneStrain(const InputParameters & parameter
                        ? getParam<unsigned int>("scalar_out_of_plane_strain_index")
-                       : 0)
+                       : 0),
+    _reference_residual_problem(dynamic_cast<ReferenceResidualProblem *>(
```
Have you had a chance to talk to @bwspenc about this?

loganharbour replied:
Disregard, see https://github.com/idaholab/moose/pull/24249#issuecomment-1550511687

loganharbour commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -44,18 +51,54 @@ class TaggingInterface
-  const std::set<TagID> & getVectorTags() const { return _vector_tags; }
+  /**
+   * Class that is used as a parameter to getVectorTags() that allows only
+   * AttribVectorTags methods to call it
```
This isn't true

loganharbour commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -120,13 +182,158 @@ class TaggingInterface
+                               Real scaling_factor);
+
+  /**
+   * Process the provided incoming residualsderivatives for the Jacobian, corresponding to the
```
```suggestion
   * Process the provided incoming residuals for the Jacobian, corresponding to the
```

loganharbour commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -173,3 +377,152 @@ class TaggingInterface
+TaggingInterface::setResidual(SystemBase & sys, const T & residual, MooseVariableFE<T> & var)
+{
+  for (const auto tag_id : _vector_tags)
+    if (sys.hasVector(tag_id))
```
I know this is old code; but since this is on your mind... there's no way we can get around this, right? Because we're lazy and we don't do the checking elsewhere?

lindsayad replied:
I feel like it's very difficult to know whether the passed in system (developer choice) will have all of the vector tags which a user may have specified in their input file (user choice). So I think this should either stay as-is or potentially be made into an exception/error

loganharbour replied:
That's fine; whatever you find appropriate. Just wanted to open the discussion while this is fresh on your mind.

lindsayad replied:
Yea it seems like we already have some things which would fail with erroring
```
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: Assertion `sys.hasVector(tag_id)' failed
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: We should not have tags in _vector_tags that do not exist in the system
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: at /home/lindad/projects/ad_kernel_abs_value_ref_23738/framework/build/header_symlinks/TaggingInterface.h, line 541
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: Stack frames: 16
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 0: libMesh::print_trace(std::ostream&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 1: void TaggingInterface::setResidual<double>(SystemBase&, double const&, MooseVariableFE<double>&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 2: NodalBC::computeResidual()
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 3: NonlinearSystemBase::computeNodalBCs(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 4: NonlinearSystemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 5: FEProblemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 6: PODFullSolveMultiApp::computeResidual()
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 7: PODFullSolveMultiApp::solveStep(double, double, bool)
stochastic_tools/test:surrogates/pod_rb/boundary.pod_boundary/train: 8: FEProblemBase::execMultiApps(MooseEnumItem, bool)
```

loganharbour commented on framework/src/interfaces/TaggingInterface.C:
```diff
@@ -157,26 +158,58 @@ TaggingInterface::useMatrixTag(TagID tag_id)
+    re_blocks.reserve(tags.size());
+    for (const auto tag_id : tags)
+    {
+      if (tags_to_skip.count(tag_id))
+        continue;
```
Did this exist before? The skipping? I think not. Seems like it would be better to filter before hand than putting this in a loop, but... I'll just let that one simmer for you. I'm okay either way.

lindsayad replied:
Changed

loganharbour reviewed (changes requested):
All minor comments.

Assembly being locked down is wonderful. I suspect this'll save us from a lot of app heartbreak over time, and will make it much easier to improve assembly.

lindsayad committed:
Protect useVectorTag and useMatrixTag

lindsayad committed:
Doxygen for reference residual problem prepareVectorTag

lindsayad committed:
Don't prepare reference vector tags if they're not in the tags containers

E.g. when preparing reference tags in the reference residual problem
context we only do so if `_vector_tags` or `_abs_vector_tags` contained
the reference tag to begin with. This was a mistake previously

GiudGiud commented on framework/include/base/Assembly.h:
```diff
@@ -768,87 +768,78 @@ class Assembly
+    friend class NonlinearSystemBase;
+    GlobalDataKey() {}
+    GlobalDataKey(const GlobalDataKey &) {}
+  };
```
what s the plan for directly modifying the petsc matrices if we re locking down all Assembly?

The tags help us do selective accumulations, do they also allow selective modifications?

lindsayad replied:
yes there are set/assignment operations

GiudGiud commented on framework/include/base/Assembly.h:
```diff
@@ -768,87 +768,78 @@ class Assembly
 
+  /**
+   * Key structure for APIs manipulating global vectors/matrices
+   */
```
please add a one-liner on how to use this.
You have to imagine the dev in the "blessed class" header seeing this being used, right clicking in VSCode and jumping to this definition

lindsayad replied:
I will imagine the dev loading their emacs tags table and jumping here 😆

GiudGiud commented on framework/src/nodalkernels/ADNodalKernel.C:
```diff
@@ -54,7 +57,10 @@ ADNodalKernel::computeJacobian()
     _qp = 0;
     const auto res = computeQpResidual();
-    _assembly.processJacobian(res, dof_idx, _matrix_tags);
+    addJacobian(_assembly,
```
how many assembly are there? I m thinking this should just live in the tagging interface

lindsayad replied:
There are very rare instances where you don't want to use `_assembly` ... displaced nodal boundary conditions is one off the top of my head

GiudGiud reviewed (commented):
pretty massive clean up. It definitely does some good to uniformize the access to assembly through the tagging interface.

look at 1/5th the files at most, @loganharbour in charge on this.

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -2818,77 +2801,152 @@ Assembly::adGradPhi<RealVectorValue>(const MooseVariableFE<RealVectorValue> & v)
-  const Real scalar = _scaling_vector ? (*_scaling_vector)(dof_index) : 1.;
-  processJacobian(residual, dof_index, matrix_tags, scalar);
-}
+  DenseVector<Number> element_vector(row_indices.size());
```
heap

lindsayad committed:
TaggingInterface doxygen fixes

lindsayad committed:
Save heap allocations

Both per Logan's suggestion in GeneralizedPlaneStrainOffDiag and
Guillaume's suggestion in Assembly

lindsayad committed:
Redesign how we handle reference residual problem

Motivated by both Logan's comment about building a set in advance
and making the reference residual API more clear for Ben

moosebuild commented:
Job [Python 3.8](https://civet.inl.gov/job/1539139/) on 89778bb : invalidated by @lindsayad

lindsayad committed:
Assembly and ReferenceResidualProblem doxy additions/fixes

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
Ok I think I've addressed all the reviews that my feeble mind understands 😄

loganharbour reviewed (approved):
LGTM.

lindsayad commented:
@GiudGiud do you want to check and see whether your conversations have been resolved?

loganharbour commented:
I'm gonna add most recipes on next; this is a big change.

moosebuild commented:
Job [Test timings](https://civet.inl.gov/job/1541016/) on 4f059d1 wanted to post the following:

View timings [here](https://mooseframework.inl.gov/docs/PRs/24206/timing)

This comment will be updated on new commits.

GiudGiud commented:
Conversations are resolved.

loganharbour commented:
Looks like the only other issue is ferret, but looks compatible. 

Things are quiet right now - I’m happy to get this in if the SAM fix is ready.

lindsayad commented:
Ok cool. Let's do it
